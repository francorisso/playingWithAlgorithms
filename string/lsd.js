export function lsd(arr, W) {
    const N = arr.length;
    const DICTIONARY = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'.split('');
    let aux = [];

    for (let letterIdx = W - 1; letterIdx >= 0; letterIdx--) {
        const count = [0];

        // count the chars
        for (let i = 0; i < N; i++) {
            const pos = DICTIONARY.indexOf(arr[i][letterIdx]);
            count[pos + 1] = (count[pos + 1] || 0) + 1;
        }

        for (let i = 1; i < count.length; i++) {
            count[i] = (count[i] || 0) + (count[i-1] || 0);
        }


        aux = [];
        for (let i = 0; i < N; i++) {
            const pos = DICTIONARY.indexOf(arr[i][letterIdx]);
            aux[ count[pos]++ ] = arr[i];
        }
        arr = aux;
    }

    return aux;
}
