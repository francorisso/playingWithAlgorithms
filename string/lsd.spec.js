import { lsd } from './lsd';

describe('lsd', () => {
    it('should sort!', () => {
        expect(lsd([
            'HOLA',
            'HOLC',
            'HOLB',
            'ADSA',
            'ADSB',
            'HZLL',
        ], 4)).toMatchSnapshot();
    });
});
