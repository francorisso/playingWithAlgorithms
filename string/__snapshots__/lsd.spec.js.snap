// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lsd should sort! 1`] = `
Array [
  "ADSA",
  "ADSB",
  "HOLA",
  "HOLB",
  "HOLC",
  "HZLL",
]
`;
